### PR TITLE
fix: improve jseditor linting performance

### LINKED ISSUE #9252: [Bug] InternalError: too much recursion
Sentry Issue: [APPSMITH-1CF](https://sentry.io/organizations/appsmith/issues/2799960887/?referrer=github_integration)

```
InternalError: too much recursion
  at getIndexOfRegex (components/editorComponents/CodeEditor/lintHelpers.ts:13:33)
  at getAllWordOccurrences (components/editorComponents/CodeEditor/lintHelpers.ts:27:11)
  at getKeyPositionInString (components/editorComponents/CodeEditor/lintHelpers.ts:41:19)
  at forEach (components/editorComponents/CodeEditor/lintHelpers.ts:94:30)
  at getLintAnnotations (components/editorComponents/CodeEditor/lintHelpers.ts:67:14)
...
(31 additional frame(s) were not displayed)
```

### LINKED ISSUE #14002: [Bug]: App crashes when content of JS Editor is very large
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

App crashes when a large JSON is pasted into JS Editor

### Steps To Reproduce

Paste a very large JSON into a js object function, App crashes.

NB: This is not reproducible on all machines

### Public Sample App

_No response_

### Version

Cloud

### PR DESCRIPTION
**Pull request by ohansFavour**
## Description

For large code editor content, the recursive process of getting the binding location during linting can cause the App to crash (depending on the size of the call stack).  This check is not necessary for a JS Object since its binding location is constant (for a valid JS Object).

## Improvements

- Performance

Difference between previous and current time to process JS Editor lint annotations (for several trials and same content).

Previous (78ms max.)

Screenshot 2022-05-21 at 15 04 47

Current (1.3 ms max.)

Screenshot 2022-05-21 at 15 01 19

## Added Feature

Custom lint error when JS Object doesn't start with 'export default'

Screenshot 2022-05-21 at 16 12 04

Screenshot 2022-05-21 at 16 13 25

Fixes #14002 
Fixes #9252

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

- Jest
- Cypress

- [x]  https://github.com/appsmithorg/TestSmith/issues/1722
- [x] https://github.com/appsmithorg/TestSmith/issues/1723
- [x] https://github.com/appsmithorg/TestSmith/issues/1724

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: fix/improve-jseditor-linting-performance 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 56.62 **(0)** | 38.64 **(0.01)** | 35.95 **(0.01)** | 56.88 **(0.01)**
 :green_circle: | app/client/src/components/editorComponents/CodeEditor/lintHelpers.ts | 94.25 **(0.5)** | 73.47 **(3.47)** | 91.67 **(1.67)** | 92.75 **(0.56)**
 :red_circle: | app/client/src/utils/WorkerUtil.ts | 88.98 **(-0.78)** | 70.59 **(-1.96)** | 100 **(0)** | 92.38 **(-0.95)**</details>

### CODE REVIEW

**Review by Rishabh-Rathod — DISMISSED** [review]
LGTM

File: app/client/src/components/editorComponents/CodeEditor/lintHelpers.ts [hunk 1]
```diff
@@ -57,15 +64,40 @@ export const getKeyPositionInString = (
...
+  isJSObject?: boolean,
 ): Annotation[] => {
   const annotations: Annotation[] = [];
   const lintErrors = errors.filter(
     (error) => error.errorType === PropertyEvaluationErrorType.LINT,
   );
   const lines = value.split("\n");
+  if (
+    isJSObject &&
+    !isEmpty(lines) &&
```
Comment (Rishabh-Rathod) [inline app/client/src/components/editorComponents/CodeEditor/lintHelpers.ts:89]: What is expected behavior when there is no code inside JSObject?
Currently, we don't show any error for that case.

Screenshot 2022-05-23 at 2 16 26 PM

Comment (ohansFavour) [reply app/client/src/components/editorComponents/CodeEditor/lintHelpers.ts:89]: When the JS Object is empty, there is a toast message informing the user that it needs to start with 'export default'. CodeMirror doesn't allow us to show lint errors on blank lines.

Comment (Rishabh-Rathod) [reply app/client/src/components/editorComponents/CodeEditor/lintHelpers.ts:89]: Understood. In that case, the changes look good to me.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/JsFunctionExecution/JSFunctionExecution_spec.ts [hunk 2]
```diff
@@ -75,4 +83,80 @@ describe("JS Function Execution", function() {
...
+        paste: true,
+        completeReplace: true,
+        toRun: false,
+        shouldCreateNewJSObj: true,
+      });
+
+      // Assert presence of toast message
+      cy.get(locator._toastMsg)
+        .should("exist")
+        .should("contain.text", invalidJSObjectStartToastMessage);
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/JsFunctionExecution/JSFunctionExecution_spec.ts:118]: Can these please be replaced with agHelper.ValidateToastMessage(invalidJSObjectStartToastMessage)

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/JsFunctionExecution/JSFunctionExecution_spec.ts [hunk 3]
```diff
@@ -75,4 +83,80 @@ describe("JS Function Execution", function() {
...
+
+      // Assert presence of lint error at the start line
+      cy.get(locator._lintErrorElement)
+        .should("exist")
+        .should("contain.text", highlightedLintText);
+    };
+
+    assertInvalidJSObjectStart(jsObjectStartingWithAComment, jsComment);
+    assertInvalidJSObjectStart(jsObjectStartingWithANewLine, jsObjectStartLine);
+    assertInvalidJSObjectStart(jsObjectStartingWithASpace, jsObjectStartLine);
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/JsFunctionExecution/JSFunctionExecution_spec.ts:128]: @ohansFavour - isnt this case contradicting with fix for [[Bug]: Unable to add comments/leave a blank line at start of JSObjects/JSEditor #12482](https://github.com/appsmithorg/appsmith/issues/12482)
So ideally after #12482 is fixed we might need to remove this case?

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/JsFunctionExecution/JSFunctionExecution_spec.ts (hunk 2 above)
Comment (Aishwarya-U-R) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/JsFunctionExecution/JSFunctionExecution_spec.ts:120]: @ohansFavour - can you please update this also?

**Pull request merged**